Content.Server/Antag/NuAntagSelectionSystem.cs

```diff
@@ -0,0 +1,365 @@
+                throw new InvalidOperationException($"Attempted to make {session} antagonist in gamerule {ToPrettyString(ent)} but there was no valid entity for player.");
+            }
+
+            antagEnt = getEntEv.Entity;
+        }
+
+        if (antagEnt is not { } player)
+            return;
+
+        //todo replace this shit down here with sloth's balling methods
+        foreach (var (_, entry) in def.Components)
+        {
+            var comp = (Component) _serialization.CreateCopy(entry.Component, notNullableOverride: true);
+            comp.Owner = player; // look im sorry ok this .owner has to live until engine api exists
+            EntityManager.RemoveComponent(player, comp.GetType());
+            EntityManager.AddComponent(player, comp);
+        }
+
+        if (session?.GetMind() is { } mind)
+        {
+            foreach (var (_, entry) in def.MindComponents)
+            {
+                var comp = (Component) _serialization.CreateCopy(entry.Component, notNullableOverride: true);
+                comp.Owner = mind; // look im sorry ok this .owner has to live until engine api exists
+                EntityManager.RemoveComponent(mind, comp.GetType());
+                EntityManager.AddComponent(mind, comp);
+            }
+
+            ent.Comp.SelectedMinds.Add((mind, Name(player)));
+        }
+
+        _inventory.SpawnItemsOnEntity(player, def.Equipment);
+
+        var afterEv = new AfterAntagEntitySelectedEvent(session, player, ent);
+        RaiseLocalEvent(ent, ref afterEv, true);
+
+        if (def.Briefing is { } briefing)
+        {
+            _antagSelection.SendBriefing(session, Loc.GetString(briefing.Text), briefing.Color, briefing.Sound);
+        }
+    }
+
+    public AntagSelectionPlayerPool GetPlayerPool(Entity<AntagSelectionComponent> ent, List<ICommonSession> sessions, AntagSelectionDefinition def)
+    {
+        var primaryList = new List<ICommonSession>();
+        var secondaryList = new List<ICommonSession>();
+        var fallbackList = new List<ICommonSession>();
+        var rawList = new List<ICommonSession>(sessions);
+        foreach (var session in sessions)
+        {
+            if (session.GetMind() is { } mind && TryComp<MindComponent>(mind, out var mindComp))
+            {
+                if (!IsSessionValid(ent, session, (mind, mindComp), def))
+                    continue;
+            }
+
+            if (!IsEntityValid(session.AttachedEntity, def))
+                continue;
+
+            var pref = (HumanoidCharacterProfile) _pref.GetPreferences(session.UserId).SelectedCharacter;
+            if (def.PrefRoles.Count == 0 || pref.AntagPreferences.Any(p => def.PrefRoles.Contains(p)))
+            {
+                primaryList.Add(session);
+            }
+            else if (def.PrefRoles.Count == 0 || pref.AntagPreferences.Any(p => def.FallbackRoles.Contains(p)))
+            {
+                secondaryList.Add(session);
+            }
+            else
```

Comment:
this should not be required, for some things you dont want it to fall back on people that havent opted in and dont want to be antags. so it should have a enum on the component or something that lets each rule customize how they want to handle it